### ISSUE TITLE
zooming metabug

### ISSUE THREAD

**Issue by jonorthwash** [jonorthwash/ud-annotatrix#143]
Current issues with graph zooming functionality:
- Autozoom default level almost always goes outside default level (#58)
- Display of graph can become too small to read with autozoom.
- No way to manually zoom (#116).
- Graph takes up much more room than it needs to at any zoom level (#51).

Please add to this list as more things come up.

My proposed solution to all of this is to implement the following:
- No more vertical or horizontal spacing between things than necessary (i.e., fix #51).
- Set a minimum limit on size of auto and manual zoom.
- Don't allow graph to grow beyond the width of the window unless it would zoom smaller than the minimum limit.
- Add a way to zoom manually (e.g., `+`/`-`/`0` keys on graph).
- Probably set a maximum size on zoom too.

Discussion welcome/encouraged.

**Comment by ftyers** [comment]
Added the keys for zooming in c424a30 and 854e434. I disabled scroll wheel zooming so that we can scroll normally.

**Comment by jonorthwash** [comment]
> I disabled scroll wheel zooming so that we can scroll normally.

Proper behaviour would preclude the need for disabling scrollwheel zooming.  E.g., if a current scroll is in process, continue the scroll and don't zoom.  But that might be too hard?

**Comment by ftyers** [comment]
Yeah, I looked into it, and it's really hard... I think this behaviour is ok for now. If we get a clamour from users we can revisit it...

**Comment by ftyers** [comment]
Can you explain what you mean with:
*  Don't allow graph to grow beyond the width of the window unless it would zoom smaller than the minimum limit.
and
* Display of graph can become too small to read with autozoom.

We've made quite a few changes to how the zoom works, perhaps they are already solved?

**Comment by jonorthwash** [comment]
> Display of graph can become too small to read with autozoom.

If a long sentence is loaded in a narrow window, things get way too small:

![2017-11-04-20 20 09_001](https://user-images.githubusercontent.com/963849/32410698-a67573c8-c19d-11e7-8093-224da15b88b6.png)

So my proposed solution is to set a minimum zoom factor for autozoom.  If the sentence can't fit in the window without going below that zoom factor, then it shouldn't—it should just wrap.

> Don't allow graph to grow beyond the width of the window unless it would zoom smaller than the minimum limit.

This was the issue where the graph would always be outside of the current view, no matter what zoom level.  It's basically fixed, but note that with narrower window sizes it likes to add about `0.5ex`'s worth of scrolling to the side (as in above screenshot).

Also, the graph should update as/after the window is resized.  Currently you have to do something else that would update it, like edit the sentence, change sentences, or reload the page.  See below for what it looks like before update:

![2017-11-04-20 23 32_001](https://user-images.githubusercontent.com/963849/32410715-0e0431f0-c19e-11e7-92ea-2add84e6a7a9.png)

**Comment by ftyers** [comment]
> So my proposed solution is to set a minimum zoom factor for autozoom. If the sentence can't fit in the window without going below that zoom factor, then it shouldn't—it should just wrap.

So you mean set the minimum default zoom to be lower ? and by "wrap" you mean scroll ?

*visualiser.js:*
```
    // Fit the graph to the window size
    cy.fit(2);
    CURRENT_ZOOM = cy.zoom(); // Get the current zoom factor.
    if(CURRENT_ZOOM >= 1.7) { // If the current zoom factor is more than 1.7, then set it to 1.7
      CURRENT_ZOOM = cy.zoom(1.7);           // This is to make sure that small trees don't appear massive.
    } else if (CURRENT_ZOOM <= 0.7) {
      CURRENT_ZOOM = cy.zoom(0.7);
    }
    cy.center();
```

> Also, the graph should update as/after the window is resized. Currently you have to do something else that would update it, like edit the sentence, change sentences, or reload the page. See below for what it looks like before update:

Ok, so the solution there is to capture window resize events and fit/center the graph after them ?

*visualiser.js:*
```
    $(window).bind('resize', onResize);
...
function onResize(e) {
    console.log('< resize event', CURRENT_ZOOM);
    cy.fit();
    cy.center();
    CURRENT_ZOOM = cy.zoom();
    console.log('> resize event', CURRENT_ZOOM);
}
```

**Comment by jonorthwash** [comment]
Did you commit this?  The behaviour has changed, but it seems buggy..

**Comment by ftyers** [comment]
Yes... and bugs??? in  _annotatrix_ ? :P

**Comment by jonorthwash** [comment]
I meant more like it doesn't seem to be implemented the way it sounded like you said it should be...

**Comment by jonorthwash** [comment]
> Ok, so the solution there is to capture window resize events and fit/center the graph after them ?

This still doesn't seem to be implemented.  When the window is resized, the graph is not resized.  Once something else is changed, it is.

**Comment by jonorthwash** [comment]
Also, when resized in and then resized back out, this happens:

![2017-11-06-11 13 49_001](https://user-images.githubusercontent.com/963849/32450954-a541be0c-c2e3-11e7-8ac8-c1d26666eea8.png)

**Comment by ftyers** [comment]
The bug is with `cy.fit()`. It doesn't behave as expected. You can test the behaviour yourself, see commit 0e29191.

**Comment by jonorthwash** [comment]
> The debug output is:

What does it think the size it's resizing to is?

What happens if we add a redraw tree call after (/before?) telling it to resize?

**Comment by ftyers** [comment]
> What does it think the size it's resizing to is?

The size of the original window.

> What happens if we add a redraw tree call after (/before?) telling it to resize?
I've tried putting the calls in different orders with the same result.

**Comment by jonorthwash** [comment]
>>   What does it think the size it's resizing to is?
>
> The size of the original window.

Ah, that could be why it isn't resizing to the newly resized size then.  Could you have it output those numbers along with the resize events?  (Because all your resize event debug lines seem to want to be the same size—so it doesn't look like it's really trying to resize...)

**Comment by ftyers** [comment]
the numbers that say "resize" are on the resize event if you look at the code.

**Comment by jonorthwash** [comment]
Also related to this: #164, #175, #191.

**Comment by jonorthwash** [comment]
Also related: #240.

**Comment by jonorthwash** [comment]
Also related: #282
